jormungandr-integration-tests/src/mock/client.rs

```diff
@@ -0,0 +1,233 @@
+        let block_ids_u8 = blocks_id
+            .iter()
+            .map(|x| x.as_ref().iter().cloned().collect())
+            .collect::<Vec<Vec<u8>>>();
+        let mut block_id = BlockIds::new();
+        block_id.set_ids(RepeatedField::from_vec(block_ids_u8));
+        self.client
+            .get_headers(grpc::RequestOptions::new(), block_id)
+    }
+
+    pub fn handshake(&self) -> HandshakeResponse {
+        let resp = self
+            .client
+            .handshake(grpc::RequestOptions::new(), HandshakeRequest::new());
+        let (_, handshake_resp, _) = resp.wait().unwrap();
+        handshake_resp
+    }
+
+    pub fn get_genesis_block_hash(&self) -> Hash {
+        Hash::from_str(&hex::encode(self.handshake().block0)).unwrap()
+    }
+
+    pub fn get_tip(&self) -> ChainHeader {
+        let resp = self
+            .client
+            .tip(grpc::RequestOptions::new(), TipRequest::new());
+        let (_, tip, _) = resp.wait().unwrap();
+        read_into(&tip.get_block_header())
+    }
+
+    pub fn upload_blocks_internal(
+        &self,
+        chain_block: ChainBlock,
+    ) -> Result<(Metadata, UploadBlocksResponse, Metadata)> {
+        let mut bytes = Vec::with_capacity(4096);
+        chain_block.serialize(&mut bytes).unwrap();
+        let mut block = Block::new();
+        block.set_content(bytes);
+        let resp = self.client.upload_blocks(
+            grpc::RequestOptions::new(),
+            grpc::StreamingRequest::single(block),
+        );
+        resp.wait()
+    }
+
+    pub fn upload_blocks(&self, chain_block: ChainBlock) -> UploadBlocksResponse {
+        let (_, response, _) = self.upload_blocks_internal(chain_block).unwrap();
+        response
+    }
+
+    pub fn upload_blocks_err(&self, chain_block: ChainBlock) -> grpc::Error {
+        self.upload_blocks_internal(chain_block).err().unwrap()
+    }
+
+    pub fn pull_blocks_to_tip(&self, from: Hash) -> Vec<ChainBlock> {
+        self.pull_blocks_to_tip_stream(from)
+            .into_future()
+            .wait_drop_metadata()
+            .unwrap()
+            .into_iter()
+            .map(|x| read_into(x.get_content()))
+            .collect()
+    }
+
+    pub fn pull_blocks_to_tip_stream(&self, from: Hash) -> grpc::StreamingResponse<Block> {
+        let mut request = PullBlocksToTipRequest::new();
+        request.set_from(RepeatedField::from_vec(vec![from.as_ref().to_vec()]));
+
+        self.client
+            .pull_blocks_to_tip(grpc::RequestOptions::new(), request)
+    }
+
+    pub fn pull_headers(&self, from: Option<Hash>, to: Option<Hash>) -> Vec<ChainHeader> {
+        self.pull_headers_stream(from, to)
+            .into_future()
+            .wait_drop_metadata()
+            .unwrap()
+            .into_iter()
+            .map(|x| read_into(x.get_content()))
+            .collect()
+    }
+
+    pub fn pull_headers_get_err(&self, from: Option<Hash>, to: Option<Hash>) -> grpc::Error {
```

Comment:
`pull_headers` could just return `Result`, so the tests can call `expect` and `expect_err` on it and you don't need two methods.